sympy/matrices/common.py

```diff
@@ -2162,13 +2161,19 @@ def entry(i, j):
             return sum(other[i,k]*self[k,j] for k in range(other.cols))
         return self._new(other.rows, self.cols, entry)
 
-    def _eval_pow_by_recursion(self, num):
+    def _eval_pow_by_recursion(self, num, mulsimp=None):
+        if mulsimp and not isinstance(mulsimp, list):
+            mulsimp = [True]*len(self)
```

Comment:
This mechanism of seems flakey. It would be better to call `dotprodsimp` in this function.

Reply:
It would be better to define a new `_eval_pow_by_recursion_mulsimp` and make `pow` link to the the customized routine, and keep the existing one, to make the existing implementation self contained for some future improvement.

Reply:
Currently the only place other than pow where `_eval_pow_by_recursion` is called is in `matrices/tests/test_matrices.py`, so if I add `_eval_pow_by_recursion_mulsimp` then `_eval_pow_by_recursion` will be present but unused, is it desirable to have dead functions sitting around?

Reply:
I'm not sure but I guess @sylee957 meant that `pow` could decide which method to call based on `mulsimp` so `_eval_pow_by_recursion_mulsimp` would only be called by `M.pow(n, mulsimp=True)`.

Reply:
No, I mean `pow(mulsimp=False, jordan=False)` should link to the `_eval_pow_by_recursion` which has no difference than the previous implementation and `pow(mulsimp=True, jordan=False)` should link to the `_eval_pow_by_recursion_mulsimp`.

I would consider these two to be a complete different matrix power algorithm, if you had introduced some more complicated contexts like masking the elements to simplify.

Reply:
Seems a little redundant but done.

Reply:
BTW, I included the `charpoly` fix with `mulsimp` flag since it is so simple so charpolys can stop blowing up.